[Proposal] Binary dependencies for SwiftPM (#1093)
apple/swift-evolution: proposals/0271-swiftpm-binary-dependencies.md

# Binary dependencies

* Proposal: [SE-271](0271-swiftpm-binary-dependencies.md)
* Authors: [Braden Scothern](https://github.com/bscothern), [Boris Bügling](https://github.com/neonichu), [Daniel Dunbar](https://github.com/ddunbar), [Franz Busch](https://github.com/FranzBusch)
* Review Manager: [Boris Bügling](https://github.com/neonichu)
* Status: **Active review (November 13...November 20)**

## Introduction

SwiftPM currently supports source-only packages for several languages, and with
a very proscriptive build model which considerably limits exactly how the
compilation of the source can be performed. While this makes packages consistent
and to some extent "simple", it limits their use in several important cases:
* Software vendors who wish to provide easy integration with the package
  manager, but do not deliver source code, cannot integrate.
* Existing code bases which would like to integrate "simply" with SwiftPM, but
  require more complicated build processes, have no recourse.

For example, consider these use cases:

>  * Someone want's to create a Swift package for

Reviewer: ```suggestion
 * Someone wants to create a Swift package for
```

   generating [LLVM](https://llvm.org) code. However, LLVM's build process is
   far more complex than can be currently fit into SwiftPM's build model. This
   makes building an *easy to use* package difficult.
 * A third-party wants to provide a Swift SDK for easily integrating their
   service with server-side Swift applications. The SDK itself relies on
   substantial amounts of internal infrastructure the company does not want to
   make available as open source.
 * A large company has an internal team which wants to deliver a Swift package
   for use in their iOS applications, but for for business reasons cannot publish
   the source code.

This proposal defines a new SwiftPM feature to allow SwiftPM to accept some

> forms of "binary packages". This proposal is intentionally written to be a

Reviewer: ```suggestion
forms of "binary packages". This proposal is intentionally written to
```

address the above use cases *explicitly*, it **does not** define a general
purpose "binary artifact" mechanism intended to address other use cases (such as
accelerating build performance). The motivations for this are discussed in more
detail below.

> Swift-evolution thread: [Discussion thread topic for that proposal](https://forums.swift.org/)

Reviewer: ```suggestion
Swift-evolution thread: [\[PITCH\] Support for binary dependencies](https://forums.swift.org/t/pitch-support-for-binary-dependencies/27620)
```

> Pitch Thread: https://forums.swift.org/t/spm-support-for-binaries-distribution/25549/24

Reviewer: Remove this line, in favour of the previous suggested change?

                        
## Motivation

SwiftPM has a large appeal to certain developer communities, like the iOS
ecosystem, where it is currently very common to rely on closed source
dependencies such as Firebase, GoogleAnalytics, Adjust and many more. Existing
package managers like Cocoapods support these use cases. By adding such support
to SwiftPM, we will unblock substantially more adoption of SwiftPM within those
communities.

Prior to Swift 5.1, the Swift compiler itself did not expose all of the features
(like ABI compatibility) required to build a workable solution. Now that those

> features are present, it makes sense to reevaluate the role of binary packages.

Reviewer: ```suggestion
features are present, it makes sense to re-evaluate the role of binary packages.
```

binary packages *where necessary*, but not to encourage their use or facilitate a
transition from the existing source-based ecosystem to a binary one.

This proposal is also focused at packages which come exclusively in binary form,
it explicitly **does not** introduce a mechanism which allows a package to be
present in either source or binary form. See alternatives considered for more
information on this choice.

## Proposed solution

To enable binary dependencies we have to make changes in the `Package.swift` manifest file. First, we propose to add a new target type which describes a binary target. Such a target needs to declare where to retrieve the artifact from and the checksum of the expected artifact. An example of such a package can be seen below:

```swift
let package = Package(
    name: "SomePackage",
    platforms: [
        .macOS(.v10_10), .iOS(.v8), .tvOS(.v9), .watchOS(.v2),
    ],
    products: [
        .library(name: "SomePackage", targets: ["SomePackageLib"])
    ],
    targets: [
        .binaryTarget(
            name: "SomePackageLib",
            artifacts: [
                .artifact(

>                     source: .url("https://github.com/some/package/releases/download/1.0.0/SomePackage-1.0.0.zip", checksum: "839F9F30DC13C30795666DD8F6FB77DD0E097B83D06954073E34FE5154481F7A")),

Reviewer: For less horizontal scrolling, add a linebreak before the `checksum:` label?

            ]
        )
     ]
)
```

Secondly we propose to add a new configuration point to the package description that allows packages to opt-out of binary dependencies. This will enforce that nothing in its transitive dependencies brings a binary dependency with it.

```swift
let package = Package(
    name: "SomeOtherPackage",
    disallowsBinaryDependencies: true,
    products: [
        .library(name: "SomeOtherPackage", targets: ["SomeOtherPackageLib"])
    ],
    targets: [
        .target(name: "SomeOtherPackageLib")
     ]
)
```

Packages are allowed to contain a mix of binary and source targets. This is
useful when, for example, providing a pre-built or closed source C library
alongside an open source set of Swift bindings for the library.

When a package is built that depends upon any product with a binary target, the
package manager will search the `artifacts` declaration list to find an artifact
which matches the current build conditions. This list
will be searched in order, and the first matching artifact will be used. It is

> the job of the package author/published to provide an appropriate set of

Reviewer: ```suggestion
the job of the package author/publisher to provide an appropriate set of
```

artifacts for the use cases the package wishes to support. This use case will be limited to Apple platforms in the beginning. In the future, we can add support for other platforms. A potential approach is outlined in the future directions section.

## Detailed design

The design consists of the following key points:
* New `PackageDescription` API for defining a binary target.
* New parameter on a package declaration level to opt-out of binary dependencies.
* New requirements for the `Package.resolved` file when using binary packages.
* A new mechanism for downloading binary target artifacts.

Terminology:

* Technically, a *target* is binary or not. However, we anticipate that often a
  single package will consist of either exclusively source or binary targets. We
  will use the term *binary package* to refer to any package which contains at
  least one binary product. Similarly, a *binary product* is one which contains
  at least one binary target.

Our design attempts to optimize for the following goals:

* Ease of use for clients

> * Easy of implementation in existing SwiftPM

Reviewer: ```suggestion
* Ease of implementation in existing SwiftPM
```

* Ease of maintenance in the face of an evolving SwiftPM
* Understandable composition with current and upcoming SwiftPM features
* Support existing well-known occurrences of binary artifacts in the existing
  (often iOS focused) target developer market.

while keeping the following as non-goals:

* Ease of production of binary packages
* Simplicity of binary artifact distribution mechanism
* Widespread use of binary packages

## New `PackageDescription` API

### BinaryTarget

> Since, a binary target is different compared to a source only target, we propose to introduce a new struct `Artifact`. This struct defines a targets associated artifacts.

Reviewer: ```suggestion
Since a binary target is different compared to a source only target, we propose to introduce a new struct `Artifact`. This struct defines a target's associated artifacts.
```

### PackageDescription
To opt out of binary packages we propose a new configuration point inside the package description.

```swift
public final class Package {
    ...
    /// This disallows any binary dependency or any transitive binary dependency.
    public var disallowsBinaryDependencies: Bool
    ...
}
```

## New `Package.resolved` Behavior

For binary targets we store the checksum of the artifact in the `Package.resolved`. This lets us check for errors during resolution where a package's version did not change but the checksum did. In this case we will throw an error alerting the user about this.

### Resolution

Package resolution and dependency expression will not be impacted by this change (except where explicitly noted).

#### Multiple references to same artifact
During resolution SwiftPM will check that all references to an artifact in a dependency graph have the same checksum.

## Binary Target Artifact Format

> SwiftPM currently supports multiple platforms; however, this proposal only proposes to add support for binary targets on Apple platforms. The reason for this is that Apple platform provide ABI guarantees and an already existing format we can leverage to make simplify the initial implementation. For Apple platforms we propose to use the `XCFramework` format for artifacts. This format already supports dynamic and static linking. Furthermore, it can contain products for every individual Apple platform at once.

Reviewer: ```suggestion
SwiftPM currently supports multiple platforms; however, this proposal only adds support for binary targets on Apple platforms. The reason for this is that Apple platforms provide ABI guarantees and an already existing format we can leverage to simplify the initial implementation. For Apple platforms we propose to use the `XCFramework` format for artifacts. This format already supports dynamic and static linking. Furthermore, it can contain products for every individual Apple platform at once.
```

Lastly, the hash of the binary is stored in the package resolved to avoid that the vendor changes the artifact behind a version without anyone noticing.

## Impact on existing packages

No current package should be affected by this change since this is only an additive change in enabling SwiftPM to use binary dependencies.

## Future directions

### Support for non-Apple platforms
Non-Apple platforms provide non-trivial challenges since they are not always giving guarantees of the ABI of the platform. Additionally, further conditions such as the corelibs-foundation ABI or if the hardware supports floating points need to be taken into consideration when declaring a package for non-Apple platforms. Various other communities tried to solve this, e.g. Python's [manylinux](https://www.python.org/dev/peps/pep-0600/).

> In the future, we could add a `ArtificatCondition` to SwiftPM which provides the possibility to declare under which conditions a certain artifact can be used. Below is a potential `ArtifactCondition` struct which does **not** include a complete set of conditions that need to be taken into consideration.

Reviewer: ```suggestion
In the future, we could add an `ArtifactCondition ` to SwiftPM which provides the possibility to declare under which conditions a certain artifact can be used. Below is a potential `ArtifactCondition` struct which does **not** include a complete set of conditions that need to be taken into consideration.
```

  user's perspective, when the source of a package is available, then one would
  typically like to think of the artifact cache as a perfect reproduction of
  "what would have been built, if I built it myself". However, leveraging a
  binary package like mechanism instead of explicit tool support for this often
  means:
  
  * There is almost no enforcement that the consumed binary artifact matches the
    source. The above presumption of equivalence makes such artifact caches a
    ripe opportunity for embedding malware into an ecosystem.

  * The consumer does not always have control over the artifact production. This
    interacts adversely with potential future SwiftPM features which would allow
    the build of a package to be more dependent on its consumer (e.g. allowing
    compile-time configuration "knobs & switches").

  * The artifact cache "optimization" may not apply to all packages, or may
    require substantial manual effort to maintain.

* When used as a workflow improvement (e.g. to reduce the scope of searches),
  our position is that the user would ultimately have a better user experience

>   by explicitly enumarting and designing features (either in SwiftPM, or in

Reviewer: ```suggestion
  by explicitly enumerating and designing features (either in SwiftPM, or in
```

  that works when no source layout is present to support binary only packages.
  
  Finally, since it would be a feature with user-authored metadata, such a
  mechanism would need to be updated when any other SwiftPM enhancement
  introduces or changes the nature of the source layout specification.

Taken together, the above points led us to focus on a proposal focused at
"vendored binaries", while our hope is that artifact caching eventually becomes
a built-in and automatic feature of the package manager which applies to all
packages.

### Binary Signatures

We considered adding signature checks during the checkout of binary dependencies but when these have transitive dependencies it gets complicated expressing that in the `Package.swift`.

```
     let package = Package(
         name: "Paper",
         products: [...],
         dependencies: [
             .package(url: "http://some/other/lib", .exact("1.2.3"), binarySignature: .any),
             .package(url: "http://some/other/lib", .exact("1.2.3"), binarySignature: .gpg("XXX")"),
         ],
         targets: [...]
     )
```

### Binary target vs. binary product

> During the discussion, it was brought up wether a binary dependency should be declared as a target or a product. Below is a list of what we took into consideration when deciding between target and product. In the end, a target seems like a better choice for a binary dependency.

Reviewer: ```suggestion
During the discussion, it was brought up whether a binary dependency should be declared as a target or a product. Below is a list of what we took into consideration when deciding between target and product. In the end, a target seems like a better choice for a binary dependency.
```

- Targets allow configuring linker/compiler flags; this ability might be necessary for static libraries
- There are already `systemLibraryTargets` which are essentially dependencies on pre-existing binaries on the host system
- Targets represent a single module, the same is true for one XCFramework
- Currently there is no way to depend on products, so mixed binary and source packages might be harder if we go with a binary product approach
- Analogy with what currently is being produced by source packages (dylibs => product)

### .o file format

> During the discussion of the proposal, the idea of using `.o` files was brought up. This would follow what SwiftPM creates for source-based dependencies right now; therefor, making the integration potentially easier. The main benefit of using `.o` files would be that the product linking the binary artifact can decide wether to link it dynamically or statically. However, further discussion needs to happen here how this would work. For now XCFrameworks are the initial format. XCFrameworks will allow current framework authors to package their existing XCFrameworks and distribute them via SwiftPM.

Reviewer: ```suggestion
During the discussion of the proposal, the idea of using `.o` files was brought up. This would follow what SwiftPM creates for source-based dependencies right now; therefore, making the integration potentially easier. The main benefit of using `.o` files would be that the product linking the binary artifact can decide whether to link it dynamically or statically. However, further discussion needs to happen here how this would work. For now XCFrameworks are the initial format. XCFrameworks will allow current framework authors to package their existing XCFrameworks and distribute them via SwiftPM.
```

(`therefor` => `therefore`; `wether` => `whether`)

### Avoiding duplicate symbols with static libraries/frameworks
When multiple products depend on a static library or framework it will result in duplicated symbols. SwiftPM could be smart enough to figure this out from the manifest and provide an error. This is a potential future improvement, but would require SwiftPM to know the linkage type of the artifact.

### Opt-in to allow binaries

> In the beginning, we considered to make binary support a opt-in feature so that only when somebody explicitly allows it then SwiftPM tried to use them. However, after discussion in the Swift forum we came to the conclusion that the trust one has to extend to a dependency is no different between a source-based and binary-based one; therefor, we made removed the opt-in behavior but added a opt-out behavior.

Reviewer: ```suggestion
In the beginning, we considered making binary support an opt-in feature so that only when somebody explicitly allows it then SwiftPM tried to use them. However, after discussion in the Swift forum we came to the conclusion that the trust one has to extend to a dependency is no different between a source-based and binary-based one; therefore, we removed the opt-in behavior but added an opt-out behavior.
```

Using an opt-in mechanism for binary dependencies would also mean that any package that adds a binary dependencies would need to do a major version bump, because it will require any client to change something in their manifests.

### Whitelist for allowed URLs for binary dependencies
During the discussion of this proposal another solution to the `allowsBinary` flag was brought up. That is to create a whitelist for URLs that are allowed origins for binary artifacts. This way one can still control from where binary dependencies come but it doesn't require to allow them for a complete dependency tree; therefore, giving more fine-grained control. However, we propose an opt-out mechanism instead. 

### Opt-out configuration in separate file

> During the discussion of this proposal it was decided that an opt-out mechanism was good to give package users and vendors an escape hatch. However, it was discussed wether this configuration should live inside the manifest or a separate configuration file. In this proposal, we opted to keep the configuration inside the manifest file.

Reviewer: ```suggestion
During the discussion of this proposal it was decided that an opt-out mechanism was good to give package users and vendors an escape hatch. However, it was discussed whether this configuration should live inside the manifest or a separate configuration file. In this proposal, we opted to keep the configuration inside the manifest file.
```

### Support for various artifact stores
Initially, we considered the various artifact stores on the market and how we can integrate with them. We decided to support a URL based artifact definition for the first implementation since the various providers require each their own method of authentication. However, we wanted to keep the possibility for future additions of providers open; therefore, we made the source of an artifact an enum which can be extended.

Possible artifact stores we considered:
- Github releases
- Github packages
- Gitlab
- Bitbucket
- Artifactory, Nexus etc.

### Resources

> The intial version of the proposal used `XCFrameworks` as the format for artifacts on Apple platforms. As a side effect that would have allowed for resources to be included via these frameworks. However, we changed the format for artifacts to `.o` files to follow current standards in SwiftPM. After all, SwiftPM itself does not support resources yet; therefore, we consider not supporting them via binary targets is fine. Once SwiftPM supports resources, binary targets should gain support for them as well.

Reviewer: Remove this outdated paragraph? (Also, typo: `intial` => `initial`)

Author: You are right does not make sense anymore with the latest changes 👍
